packages/rest/src/events/WebSocketEvents.ts

```diff
@@ -0,0 +1,95 @@
+import type { ServerConfig } from '../utils/ServerConfig'
+import type {
+  Agent,
+  BasicMessageStateChangedEvent,
+  ConnectionStateChangedEvent,
+  CredentialStateChangedEvent,
+  ProofStateChangedEvent,
+} from '@aries-framework/core'
+
+import {
+  BasicMessageEventTypes,
+  ConnectionEventTypes,
+  CredentialEventTypes,
+  ProofEventTypes,
+} from '@aries-framework/core'
+import WebSocket from 'ws'
+
+import { sendWebhookEvent } from './WebhookEvent'
+
+export const webSocketEvents = async (
+  server: WebSocket.Server,
+  stream: WebSocket,
+  agent: Agent,
+  config: ServerConfig
+) => {
+  stream.on('message', function message(data) {
+    server.clients.forEach((client) => {
+      if (client.readyState === WebSocket.OPEN) {
+        client.send(data)
+      }
+    })
+  })
+
+  agent.events.on(
+    BasicMessageEventTypes.BasicMessageStateChanged,
+    async ({ payload }: BasicMessageStateChangedEvent) => {
+      const record = payload.basicMessageRecord
+      const body = record.toJSON()
+
+      if (config.webhookUrl) {
+        await sendWebhookEvent(config.webhookUrl + '/basic-messages', body, agent.config.logger)
+      }
+
+      server.clients.forEach((client) => {
+        if (client.readyState === WebSocket.OPEN) {
+          client.send(body)
+        } else if (client.readyState === WebSocket.CLOSED) {
+          client.terminate()
+        }
+      })
```

Comment:
Maybe we can generalize this code a bit? We have the sendWebhookEvent method already, what if we create a single method called `emitEventToClient` that will a) send the webhook event to the specified url and b) send a websocket event over the existing websocket clients.

This way the type of the event, and emitting the event to the client is a bit more separated